Encryption.md

```diff
@@ -19,216 +19,544 @@
+128, 192 and 256 bits. 
 
+Initially, two algorithms have been implemented, one based on a GCM mode of AES, and the 
+other on a combination of GCM and CTR modes.
 
-## File Format
+AES GCM is an authenticated encryption. Besides the data confidentiality (encryption), it 
+supports two levels of integrity verification (authentication): of the data (default), 
+and of the data combined with an optional AAD (“additional authenticated data”). The 
+default authentication allows to make sure the data has not been tampered with. An AAD 
+is a free text to be signed, together with the data. The user can, for example, pass the 
+file name with its version (or creation timestamp) as an AAD input, to verify that the 
+file has not been replaced with an older version. The details on how Parquet creates 
+and uses AADs are provided in the section 4.4.
 
-The encrypted Parquet files have a different extension - “.parquet.encrypted”.
-
-The encryption is flexible - each column and the footer can be encrypted with the same key, 
-with a different key, or not encrypted at all.
-
-The metadata structures (`PageHeader`, `ColumnIndex`, `OffsetIndex`; and sometimes `FileMetaData` and 
-`ColumnMetaData`, see below) are encrypted after Thrift serialization. For each structure, 
-the encryption buffer is comprised of an IV, ciphertext and tag, as described in the 
-Algorithms section. The length of the encryption buffer (a 4-byte little endian) is 
-written to the output stream, followed by the buffer itself.
-
-The column pages (data and dictionary) are encrypted after compression. For each page, 
-the encryption buffer is comprised of an IV, ciphertext and (in case of AES_GCM_V1) of a 
-tag, as described in the Algorithms section. Only the buffer is written to the output 
-stream - not need to write the length of the buffer, since the length (size of the page after
-compression and encryption) is kept in the page headers.
-
-A `crypto_meta_data` field in set in each `ColumnChunk` in the encrypted columns. 
-`ColumnCryptoMetaData` is a union - the actual structure is chosen depending on whether the 
-column is encrypted with the footer key, or with a column-specific key. For the latter, 
-a key metadata can be specified, with a maximal length of 512. Key metadata is a free-form
-byte array that can be used by a reader to retrieve the column encryption key. 
-
-Parquet file footer, and its nested structures, contain sensitive information - ranging 
+Sometimes, a hardware acceleration of AES is unavailable (e.g. in Java 8). Then AES crypto 
+operations are implemented in software, and can be somewhat slow, becoming a performance 
+bottleneck in certain workloads. AES CTR is a regular (not authenticated) cipher. It is 
+faster than GCM cipher, since it doesn’t perform integrity verification and doesn’t 
+calculate the authentication tag. Actually, GCM is a combination of CTR cipher and an 
```

Comment:
The term "authentication tag" has not been defined. Can you add it to the description of AES-GCM above?